### PR TITLE
Practice app/auth0

### PR DESCRIPTION
**Pull request by ersoykadir**
PR finished jobs:
- Users backend tests done, 
- Users backend login-register-getUsername done,
- Users frontend, login-register done,
- main page frontend for practice app done.

I have implemented the user login and register for the practice-app, and also authorization.

Flow:
- User registers to system with email, name, surname and password.
- Backend creates a user object in db with given credentials so that they can login.
- User logins to the system with credentials that they provided while registering.
- Backend provides a auth token behind the scenes so that user can view pages that authorization is necessary.
- User is redirected to main page where there is multiple buttons to use.
- Each button redirects user to a feature of our practice app, 
  - Ex: Chess button redirects user to chess page where user can start a chess game against an AI and play.
### Screenshots
image
image
image

### CODE REVIEW

File: practice-app/client/pages/user/main.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            toast.warning("Something went wrong. Try again.");
+        }
+    };
+    useEffect(() => {
+        getUsername();
+    }, []);
+    return (
+        <>
+            <h1>Welcome to The Desert of the Real ! </h1>
+                <h1>{user_name}</h1>
```
Comment (nurlandadashov02) [inline practice-app/client/pages/user/main.js:35]: Suggestion: Can you please make username h2 and smaller?

File: practice-app/client/pages/user/main.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            {/* <Image src="/desert.png"  width="729" height="300" /> */}
+            <div className='buttons'>
+                {/* <div>
+                    {Endpoints.map(MainButton, this)}
+                </div> */}
```
Comment (nurlandadashov02) [inline practice-app/client/pages/user/main.js:40]: remove commented lines.

Comment (nurlandadashov02) [inline practice-app/client/pages/user/main.js:40]: remove commented lines.

File: practice-app/client/pages/user/main.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+                <MainButton type="submit" onClick={handleClick}>Chess</MainButton>
+                <MainButton type="submit">Login</MainButton>
+                <MainButton type="submit">Login</MainButton>
+                <MainButton type="submit">Login</MainButton>
+                <MainButton type="submit">Login</MainButton>
+                <MainButton type="submit">Login</MainButton>
```
Comment (nurlandadashov02) [inline practice-app/client/pages/user/main.js:47]: Buttons should not have types. type="submit" is usually used within forms.

Comment (nurlandadashov02) [inline practice-app/client/pages/user/main.js:47]: Buttons should not have types. type="submit" is usually used within forms.

File: practice-app/server/src/routes/users/users.controller.js [hunk 4]
```diff
@@ -83,29 +83,49 @@ const UserController = {
...
-                return res.status(200).json({
-                    access_token: response.data.access_token,
-                })
-            }
-            else{
-                return res.status(400).json({
-                    message:" Failed to acquire access token!",
-                });
+            const user = await UserModel.getUserByEmail(email);
+            console.log(user)
```
Comment (nurlandadashov02) [inline practice-app/server/src/routes/users/users.controller.js:87]: remove console.log

Comment (nurlandadashov02) [inline practice-app/server/src/routes/users/users.controller.js:87]: remove console.log

File: practice-app/server/src/tests/users/users.test.js [hunk 5]
```diff
@@ -0,0 +1,292 @@
+const supertest = require("supertest");
+const { MongoMemoryServer } = require("mongodb-memory-server-core");
+const mongoose = require("mongoose");
+const app = require("../../app");
+const axios = require("axios");
+const ChessGame = require("../../models/chess/chess.model");
+const {User} = require("../../models/users/users.model");
+const { dbConnect, dbDisconnect } = require("../utils/db");
+jest.mock("axios");
```
Comment (nurlandadashov02) [inline practice-app/server/src/tests/users/users.test.js:9]: remove unused imports

Comment (nurlandadashov02) [inline practice-app/server/src/tests/users/users.test.js:9]: remove unused imports

File: practice-app/server/src/tests/users/users.test.js [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                axios.post.mockResolvedValueOnce({
+                    data: {
+                        access_token: "access_token",
+                    },
+                });
+                axios.post.mockResolvedValueOnce({
+                    data: {
+                        created_at: "created_at",
+                    },
+                });
```
Comment (nurlandadashov02) [inline practice-app/server/src/tests/users/users.test.js:113]: why are you mocking axios.post twice?

Comment (nurlandadashov02) [inline practice-app/server/src/tests/users/users.test.js:113]: why are you mocking axios.post twice?

File: practice-app/server/src/tests/users/users.test.js [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                    response: {
+                        status: 409,
+                        data:{
+                            statusCode: 409,
+                            error: "Conflict",
+                            message: 'The user already exists.',
+                            errorCode: 'auth0_idp_error' 
+                        }
+                    }
+                });
```
Comment (nurlandadashov02) [inline practice-app/server/src/tests/users/users.test.js:150]: Why are there two mocks for axios.post

Comment (nurlandadashov02) [inline practice-app/server/src/tests/users/users.test.js:150]: Why are there two mocks for axios.post

File: practice-app/server/src/routes/users/users.route.js [hunk 8]
```diff
@@ -19,7 +19,7 @@ const usersRouter = express.Router();
  *     parameters:
  *      - in: "body"
  *        name: "Body"
- *        description: "descc"
+ *        description: "Creates a user with given data."
  *        required: true
  *        schema:
  *          type: object
```
Comment (nurlandadashov02) [inline practice-app/server/src/routes/users/users.route.js:25]: Your endpoints do not work on swagger. Please remove parameters and change it to requestBody as I explained.

Comment (nurlandadashov02) [inline practice-app/server/src/routes/users/users.route.js:25]: Your endpoints do not work on swagger. Please remove parameters and change it to requestBody as I explained.

File: practice-app/server/src/routes/users/users.controller.js [hunk 9]
```diff
@@ -83,29 +83,49 @@ const UserController = {
...
         } catch (error) {
             return res.status(400).json({
-                message: error.toString(),
+                message:"Failed to acquire access token!",
             });
         }
     },
+    getUsername: async function (req, res) {
+        let email = req.query.email;
+        console.log(email)
```
Comment (nurlandadashov02) [inline practice-app/server/src/routes/users/users.controller.js:106]: remove console.log

Comment (nurlandadashov02) [inline practice-app/server/src/routes/users/users.controller.js:106]: remove console.log

**Review by nurlandadashov02 — CHANGES_REQUESTED** [review]
Great job Kadir. Only a few minor changes are needed.

File: practice-app/server/src/tests/users/users.test.js (hunk 7 above)
Comment (ersoykadir) [reply practice-app/server/src/tests/users/users.test.js:150]: There is two axios request for register endpoint. One for admin token and one for creating user.

File: practice-app/server/src/routes/users/users.route.js (hunk 8 above)
Comment (ersoykadir) [reply practice-app/server/src/routes/users/users.route.js]: Swagger definitions updated, They work now.

Comment (ersoykadir) [reply practice-app/server/src/routes/users/users.route.js]: Swagger definitions updated, They work now.

File: practice-app/server/src/routes/users/users.controller.js (hunk 9 above)
Comment (ersoykadir) [reply practice-app/server/src/routes/users/users.controller.js:106]: Done!

Comment (ersoykadir) [reply practice-app/server/src/routes/users/users.controller.js:106]: Done!

File: practice-app/client/pages/user/main.js [hunk 10]
```diff
@@ -11,7 +11,11 @@ export default function main() {
     const handleClick = async (values) => {
         router.push(`/chess`);
     };
-
+    if (typeof window !== 'undefined') {
+        if(!localStorage.getItem("access_token")){
+            router.push(`/user/login`);
+        };
+    }
```
Comment (nurlandadashov02) [inline practice-app/client/pages/user/main.js:18]: I guess you want to redirect user if there is not access token. The correct place to do this is useEffect. You will not `if (typeof window !== 'undefined') {` there.

**Review by nurlandadashov02 — APPROVED** [review]

**Pull request merged**